### ISSUE TITLE
Reusing Workers in Emscripten

### ISSUE THREAD

**Issue by allsey87** [emscripten-core/emscripten#16549]
I need to get a (not so well designed) application running under Emscripten. I say "not so well designed" since the application starts up a large number of threads, most of which spend their time sleeping. This is an issue since, as I understand, Emscripten currently needs a worker for each active thread and the number of available workers is limited in the browser.

I see that threads started with `canBlock` set to false, implement `emscripten_futex_wait` by calling out to `_emscripten_futex_wait_non_blocking` and busy-waiting in Javascript. I am curious to know whether it is possible change the implementation of `__emscripten_futex_wait_non_blocking` to something similar to below, where the stack is saved and the worker is free to start executing a different pthread:

```javascript
function __emscripten_futex_wait_non_blocking(addr, val, timeout) {
  withStackSave(function() {
   var varargs = stackAlloc(12);
   var tNow = performance.now();
   var tEnd = tNow + timeout;
   GROWABLE_HEAP_I32()[varargs >> 2] = addr;
   GROWABLE_HEAP_I32()[varargs + 4 >> 2] = val;
   GROWABLE_HEAP_F32()[varargs + 8 >> 2] = tEnd;
   enqueue_waiting_thread();
  });
 }
```
Where `enqueue_waiting_thread` puts pthreads into a list which is routinely checked by the runtime until their futex is no-longer contended and then reschedules them to be executed by an available worker. Is this a terrible idea or are there any reasons why something like this wouldn't work?

**Comment by sbc100** [comment]
What you are suggesting is some kind of green threading (cooperative user-space threading).   We don't have support for that today, and it would be quite a lot of work to support many-to-one worker-to-pthread relationship like this. 

A couple of side notes: 

- We do have a fibre API that allows cooperative task switching within a pthread: https://emscripten.org/docs/api_reference/fiber.h.html.    Its does rely on ASYNCIFY though.  
- There is a project called GNU PTh which is an implementation of pthreads that uses cooperative scheduling and switching using setcontext/getcontext (which emscripten doesn't support today but could be implemented using ASYNCIFY/fibres): https://www.gnu.org/software/pth/
- As of today all worker set `canBlock` to true.. Its only the main browser thread that cannot block.

**Comment by sbc100** [comment]
Also, note that the code for futex wait is all now in native code:

https://github.com/emscripten-core/emscripten/blob/500d8267b4eb22fcc6606aa695a1a076c0f40375/system/lib/pthread/emscripten_futex_wait.c#L111

**Comment by allsey87** [comment]
Just to be clear, would using `stackSave` and `stackRestore` inside `__emscripten_futex_wait_non_blocking` cause problems? I am working with a custom runtime and the application in question has between compiled against Emscripten 3.0.1

**Comment by sbc100** [comment]
In general its should be fine to call `stackSave` and `stackRestore` anywhere.  There are some caveats to that in the each pthread is allocated its own stack on startup and so if you use any features that want to check that the `__stack_point` is in the correct range (for example `STACK_OVERFLOW_CHECK`) then you will want to make sure you don't set it to a value that is outside the current min/max.

See https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/stack_limits.S
And https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/stack_ops.S

Also there is bunch of other state that such as the thread id (pthread_self()) that would need to be switched too if you want to do full green threads: https://github.com/emscripten-core/emscripten/blob/main/system/lib/pthread/emscripten_thread_state.S

**Comment by allsey87** [comment]
Right, I see the main source of my confusion... I thought `stackSave` and `stackRestore` were actually unwind/rewinding the stack. Perhaps a better name for these functions would be `getStackPointer` and `setStackPointer`?

**Comment by juj** [comment]
>  I am curious to know whether it is possible change the implementation of __emscripten_futex_wait_non_blocking to something ... where the stack is saved and the worker is free to start executing a different pthread:

There is a misunderstanding here about what the WebAssembly/Emscripten observable stack actually contains. Unlike native platforms, the stack does not contain any information about control flow. Emscripten observable stack only contains "spillover" data variables that were not possible to be made part of the secure Wasm VM stack. (because they were big arrays or had their addresses taken)

The secure Wasm VM stack itself contains all the control flow, as well as most function local variables, and all of that information is hidden from being manipulated in any way by C and/or JS code.

So if you call into `__emscripten_futex_wait_non_blocking` and then switched the stack underneath, it would not cause the call control flow stack that led the program into `__emscripten_futex_wait_non_blocking` function to get switched underneath it, to start executing/return to some other function. Instead it would just mess up the spillover data that LLVM had decided to push to the external WebAssembly/Emscripten data stack.

**Comment by allsey87** [comment]
Thanks for the input @juj and @sbc100, I will close this issue since it doesn't seem possible to implement what I was suggesting. At least not without pretty serious changes to Emscripten.
